src/svm/svr.rs

```diff
@@ -0,0 +1,538 @@
+            let sj = T::two() * T::from_usize(j).unwrap() - T::one();
+            for v in self.sv.iter_mut() {
+                v.grad[0] -= si * k1[v.index] * delta_alpha_i + sj * k2[v.index] * delta_alpha_j;
+                v.grad[1] += si * k1[v.index] * delta_alpha_i + sj * k2[v.index] * delta_alpha_j;
+            }
+
+            self.minmax();
+        }
+
+        let b = -(self.gmax + self.gmin) / T::two();
+
+        let mut result: Vec<M::RowVector> = Vec::new();
+        let mut alpha: Vec<T> = Vec::new();
+
+        for v in self.sv {
+            if v.alpha[0] != v.alpha[1] {
+                result.push(v.x);
+                alpha.push(v.alpha[1] - v.alpha[0]);
+            }
+        }
+
+        (result, alpha, b)
+    }
+}
+
+impl<T: Clone> Cache<T> {
+    fn new(n: usize) -> Cache<T> {
+        Cache {
+            data: vec![RefCell::new(None); n],
+        }
+    }
+
+    fn get<F: Fn() -> Vec<T>>(&self, i: usize, or: F) -> Ref<Vec<T>> {
+        if self.data[i].borrow().is_none() {
+            self.data[i].replace(Some(or()));
+        }
+        Ref::map(self.data[i].borrow(), |v| v.as_ref().unwrap())
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::linalg::naive::dense_matrix::*;
+    use crate::metrics::mean_squared_error;
+    use crate::svm::*;
+
+    #[test]
+    fn svr_fit_predict() {
+        let x = DenseMatrix::from_2d_array(&[
+            &[234.289, 235.6, 159.0, 107.608, 1947., 60.323],
+            &[259.426, 232.5, 145.6, 108.632, 1948., 61.122],
+            &[258.054, 368.2, 161.6, 109.773, 1949., 60.171],
+            &[284.599, 335.1, 165.0, 110.929, 1950., 61.187],
+            &[328.975, 209.9, 309.9, 112.075, 1951., 63.221],
+            &[346.999, 193.2, 359.4, 113.270, 1952., 63.639],
+            &[365.385, 187.0, 354.7, 115.094, 1953., 64.989],
+            &[363.112, 357.8, 335.0, 116.219, 1954., 63.761],
+            &[397.469, 290.4, 304.8, 117.388, 1955., 66.019],
+            &[419.180, 282.2, 285.7, 118.734, 1956., 67.857],
+            &[442.769, 293.6, 279.8, 120.445, 1957., 68.169],
+            &[444.546, 468.1, 263.7, 121.950, 1958., 66.513],
+            &[482.704, 381.3, 255.2, 123.366, 1959., 68.655],
+            &[502.601, 393.1, 251.4, 125.368, 1960., 69.564],
+            &[518.173, 480.6, 257.2, 127.852, 1961., 69.331],
+            &[554.894, 400.7, 282.7, 130.081, 1962., 70.551],
+        ]);
+
+        let y: Vec<f64> = vec![
+            83.0, 88.5, 88.2, 89.5, 96.2, 98.1, 99.0, 100.0, 101.2, 104.6, 108.4, 110.8, 112.6,
+            114.2, 115.7, 116.9,
+        ];
+
+        let y_hat = SVR::fit(
+            &x,
+            &y,
+            LinearKernel {},
+            SVRParameters {
+                eps: 2.0,
+                c: 10.0,
+                tol: 1e-3,
+            },
+        )
+        .and_then(|lr| lr.predict(&x))
+        .unwrap();
+
+        println!("{:?}", y_hat);
```

Comment:
do we want to have print statements in tests?

Reply:
thanks for spotting it!